cpp/src/strings/convert/convert_datetime.cu

```diff
@@ -427,9 +424,166 @@ std::unique_ptr<cudf::column> to_timestamps(strings_column_view const& strings,
+ *
+ * This does no data conversion.
+ */
+struct check_datetime_format {
+  column_device_view const d_strings;
+  format_item const* d_format_items;
+  size_type items_count;
+
+  /**
+   * @brief Check the specified characters are between ['0','9'].
+   *
+   * @param str Beginning of characters to check.
+   * @param bytes Number of bytes to check.
+   * @return true if all digits are 0-9
+   */
+  __device__ bool check_digits(const char* str, size_type bytes)
+  {
+    return thrust::all_of(thrust::seq, str, str + bytes, [] __device__(char chr) {
+      return (chr >= '0' && chr <= '9');
+    });
+  }
+
+  /**
+   * @brief Check the specified characters are between ['0','9']
+   * and the resulting integer is within [`min_value`, `max_value`].
+   *
+   * @param str Beginning of characters to check.
+   * @param bytes Number of bytes to check.
+   * @param min_value Inclusive minimum value
+   * @param max_value Inclusive maximum value
+   * @return true if parsed value is between `min_value` and `max_value`.
+   */
+  __device__ bool check_value(const char* str, size_type bytes, int min_value, int max_value)
+  {
+    const char* ptr = str;
+    int32_t value   = 0;
+    for (size_type idx = 0; idx < bytes; ++idx) {
+      char chr = *ptr++;
+      if (chr < '0' || chr > '9') return false;
+      value = (value * 10) + static_cast<int32_t>(chr - '0');
+    }
+    return value >= min_value && value <= max_value;
+  }
+
+  /**
+   * @brief Check the string matches the format.
+   *
+   * Walk the `format_items` as we read the string characters
+   * checking the characters are valid for each format specifier.
+   * The checking here is a little more strict than the actual
+   * parser used for conversion.
+   */
+  __device__ bool check_string(string_view const& d_string)
+  {
+    auto ptr    = d_string.data();
+    auto length = d_string.size_bytes();
+    for (size_t idx = 0; idx < items_count; ++idx) {
+      auto item = d_format_items[idx];
+      // eliminate static character values first
+      if (item.item_type == format_char_type::literal) {
+        // check static character matches
+        if (*ptr != item.value) return false;
+        ptr += item.length;
+        length -= item.length;
+        continue;
+      }
+      // allow for specifiers to be truncated
+      if (item.value != 'f')
+        item.length = static_cast<int8_t>(std::min(static_cast<size_type>(item.length), length));
+
+      // special logic for each specifier
+      // reference: https://man7.org/linux/man-pages/man3/strptime.3.html
+      bool result = false;
+      switch (item.value) {
+        case 'Y': result = check_digits(ptr, item.length); break;
+        case 'y': result = check_digits(ptr, item.length); break;
+        case 'm': result = check_value(ptr, item.length, 1, 12); break;
+        case 'd': result = check_value(ptr, item.length, 1, 31); break;
+        case 'j': result = check_value(ptr, item.length, 1, 366); break;
```

Comment:
nit: doesn't seem to check if 366 is in actual leap year. Maybe that's too complex and not requested.